diff -Naur linux.orig/include/linux/console_struct.h linux.new/include/linux/console_struct.h
--- linux.orig/include/linux/console_struct.h	2009-06-24 23:28:53.000000000 +0000
+++ linux.new/include/linux/console_struct.h	2009-06-24 23:30:02.000000000 +0000
@@ -131,7 +131,7 @@
 #define CUR_HWMASK	0x0f
 #define CUR_SWMASK	0xfff0

-#define CUR_DEFAULT CUR_UNDERLINE
+#define CUR_DEFAULT CUR_NONE

 #define CON_IS_VISIBLE(conp) (*conp->vc_display_fg == conp)

